// react-native-swipeable-flat-list/src/components/SwipeableFlatList.js
'use strict';

import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { FlatList, PanResponder, View } from 'react-native';

import SwipeableListItem from './SwipeableListItem';


export default class SwipeableFlatList extends Component {
	static propTypes = {
		data: PropTypes.arrayOf(PropTypes.object).isRequired,
		renderItem: PropTypes.func.isRequired,
		renderLeft: PropTypes.func,
		renderRight: PropTypes.func,
		itemBackgroundColor: PropTypes.string,
	};

	static defaultProps = {
		backgroundColor: '#fff',
	};

	constructor(props) {
		super(props);
		this._itemRef = React.createRef();
	}

	_handleOpenChild (ref,index) {
		this._itemRef?.current?.close?.();
		this._itemRef = ref;
		this.props.onopen(index)
	}

	_handleCloseChild (ref,index)  {
		this._itemRef = undefined;
		this.props.onclose(index)
	}

	_handleScroll = (event) => {
		const { onScroll } = this.props;
		this._itemRef?.close?.();
		onScroll?.(event);
	}

	render = () => {
		const {
			data,
			itemBackgroundColor,
			onScroll,
			renderItem,
			renderLeft,
			renderRight,
			...otherProps
		} = this.props;
		return (
			<FlatList
				{...otherProps}
				data={data}
				
				renderItem={({ item,index }) =>
					<SwipeableListItem
					isSwiping={isSwiping => this.props.isSwiping(isSwiping)}
						item={renderItem({ item ,index})}
						left={renderLeft && renderLeft({ item })}
						right={renderRight && renderRight({ item })}
						backgroundColor={itemBackgroundColor}
						onOpen={(ref)=>this._handleOpenChild  (ref,index)}
						onClose={(ref)=>this._handleCloseChild(ref,index)}
					/>
				}
				onScroll={this._handleScroll}
			/>
		);
	}
}
